### PR TITLE
Finalization for scalog

### PR DESCRIPTION
**Pull request by evantzhao**
This code touches the data and ordering layer. We assume that the kubernetes operator will be able to send a signal to the ordering layer, so the finalization process starts from there. Ordering should send a signal to the data layer to tell it to kill its communication channels and to stop taking append reqeusts from clients. Both layers should clean up metadata used in communication.

### CODE REVIEW

**Comment by ccding** [comment]
three minor issues

(1) allow each finalize request to include multiple shards (in analytics jobs, one may want to finalize a lot of shards at some point; it's just too expensive to do them one by one)

(2) can we make a configurable k, say a shard has to be finalized after k cuts (see section 4.4)

(3) finalize request should be committed by Raft and totally ordered with the cuts

File: order/order_server.go [hunk 1]
```diff
@@ -153,6 +153,33 @@ func (server *orderServer) updateCommittedCutGlobalSeqNumAndBroadcastDeltas(delt
 	}
 }
 
+/*
+notifyShardAndDeleteShardMetadata sends a finalization message to each replica
+within [shardID], and removes the remaining shardID metadata from the ordering layer.
+
+This operation acquires a writer lock
+*/
+func (server *orderServer) notifyShardAndDeleteShardMetadata(shardID int) {
```
Comment (evantzhao) [inline order/order_server.go]: @davidchuyaya Could you double check that I'm properly guarding structures like commitedGlobalCut and contestedGlobalCut? You probably know more about what raft is doing in the background than I am and its always better if we have more eyeballs on synchronization logic! Also I hope I cleaned up all the ds associated with a shardID lol

**Comment by evantzhao** [comment]
@ccding What is the motivation behind finalizing after `k` more cuts? Just curious

File: order/order_server.go (hunk 1 above)
Comment (davidchuyaya) [reply order/order_server.go]: I made some changes, should be good.

File: order/main.go [hunk 2]
```diff
@@ -36,7 +37,7 @@ func Start() {
 	id, peers := getRaftIndexPeerUrls()
 	// TODO: remove hard coded server shard count
 	// TODO: Allow new shards to join the ordering layer
-	server = newOrderServer(make([]int, 1, 1), 2, nil, nil)
+	server = newOrderServer(golib.NewSet(1, 1), 2, nil, nil)
```
Comment (evantzhao) [inline order/main.go]: Can this actually be started with an initial size of zero? I'll worry about passing in the proper number of servers per shard

Comment (davidchuyaya) [reply order/main.go]: Yup that will also work

**Comment by ccding** [comment]
> @ccding What is the motivation behind finalizing after k more cuts? Just curious

If we want to finalize a shard due to some maintenance schedule, we want to give clients extra time to prepare to connect to other shards. Once a storage server receives "finalize after $k$ cuts", it informs clients but continues accepting requests. It is expected that after $k$ cuts, most of the clients (if not all) will be moved to other shards. It reduces hiccups during the process.

**Comment by ccding** [comment]
> @ccding What is the motivation behind finalizing after k more cuts? Just curious

If we are going to finalize a shard due to some scheduled maintenance, we want to give clients extra time to prepare to connect to other shards. Once a storage server receives "finalize after k cuts", it informs clients but continues accepting requests until the end of the k cuts. It is expected that after k cuts, most of the clients (if not all) will be moved to other shards. This reduces hiccups during the process.

**Pull request merged**